File: src/Clueless/GameBoard.m
//
//  GameBoard.m
//  Clueless
//
//  Created by John Gilday on 11/23/12.
//  Copyright (c) 2012 Johns Hopkins University. All rights reserved.
//

#import "GameBoard.h"
#import "GameBoardSpace.h"
#import "Room.h"
#import "HallwaySpace.h"
#import "ClueCharacter.h"

@implementation GameBoard

@synthesize spaces;
@synthesize players;

-(id) initWithPlayers: (NSArray*) pplayers
{
    self = [super init];
    
    [self createSpaces];
    self.players = pplayers;
    
    return self;
}

-(NSArray*) getPossibleMoves:(Player *)player
{
    GameBoardSpace *currentSpace = [spaces objectForKey:[player location]];
    NSMutableArray *possibleSpaces = [[NSMutableArray alloc] init];
    for(GameBoardSpace *space in currentSpace.navigationTargets)
    {
        if([space isMemberOfClass:[HallwaySpace class]])
        {
            BOOL hallwaySpaceIsOccupied = NO;
            for (Player *otherPlayer in players)
            {
                GameBoardSpace *otherPlayerPosition = [spaces objectForKey:[otherPlayer location]];
                hallwaySpaceIsOccupied = [otherPlayerPosition.spaceId isEqualToString:space.spaceId];
                if(hallwaySpaceIsOccupied) break;
            }
            if(!hallwaySpaceIsOccupied) [possibleSpaces addObject:space];
        } else
        {
            [possibleSpaces addObject:space];
        }
    }
    return possibleSpaces;
}

-(BOOL) isPlayerAbleToMove:(Player *)player toSpace: (NSString *) spaceId
{
    NSArray* possibleSpaces = [self getPossibleMoves:player];
    
    for(GameBoardSpace *space in possibleSpaces)
    {
        NSLog(@"Can player at %@ move to %@?", player.location, space.spaceId);
        if([space.spaceId isEqualToString:spaceId]) return YES;
    }
    return NO;
}

-(void) createSpaces
{
    spaces = [[NSMutableDictionary alloc] init];
    
    Room *study = [[Room alloc] initWithId: Study];
    Room *hall = [[Room alloc] initWithId: Hall];
    HallwaySpace *hallway = [[HallwaySpace alloc] initWithRoomsToConnect:study ConnectsTo:hall];
    [self addGameBoardSpace:study];
    [self addGameBoardSpace:hall];
    [self addGameBoardSpace:hallway];
    
    Room *lounge = [[Room alloc] initWithId: Lounge];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:hall ConnectsTo:lounge];
    [self addGameBoardSpace:lounge];
    [self addGameBoardSpace:hallway];
    
    Room *diningRoom = [[Room alloc] initWithId: DiningRoom];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:lounge ConnectsTo:diningRoom];
    [self addGameBoardSpace:diningRoom];
    [self addGameBoardSpace:hallway];

    Room *kitchen = [[Room alloc] initWithId: Kitchen];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:kitchen ConnectsTo:diningRoom];
    [self addGameBoardSpace:kitchen];
    [self addGameBoardSpace:hallway];
    
    Room *ballRoom = [[Room alloc] initWithId:BallRoom];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:ballRoom ConnectsTo:kitchen];
    [self addGameBoardSpace:ballRoom];
    [self addGameBoardSpace:hallway];

    Room *conservatory = [[Room alloc] initWithId:Conservatory];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:conservatory ConnectsTo:ballRoom];
    [self addGameBoardSpace:conservatory];
    [self addGameBoardSpace:hallway];

    Room *library = [[Room alloc] initWithId:Library];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:library ConnectsTo:conservatory];
    [self addGameBoardSpace:library];
    [self addGameBoardSpace:hallway];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:library ConnectsTo:study];
    [self addGameBoardSpace:hallway];
    
    // Add billiard room in the middle
    Room *billiardRoom = [[Room alloc] initWithId: BilliardRoom];
    [self addGameBoardSpace:billiardRoom];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:hall ConnectsTo:billiardRoom];
    [self addGameBoardSpace:hallway];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:billiardRoom ConnectsTo:library];
    [self addGameBoardSpace:hallway];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:ballRoom ConnectsTo:billiardRoom];
    [self addGameBoardSpace:hallway];
    hallway = [[HallwaySpace alloc] initWithRoomsToConnect:diningRoom ConnectsTo:billiardRoom];
    [self addGameBoardSpace:hallway];
    
    // Create secret passages
    [kitchen connectSpace:study];
    [conservatory connectSpace:lounge];
    
    // Create start spots
    GameBoardSpace *msScarlet = [[GameBoardSpace alloc] initWithId:MsScarletID];
    [[msScarlet navigationTargets] addObject:[spaces valueForKey:@"Hall-Lounge"]];
    [self addGameBoardSpace:msScarlet];
    
    GameBoardSpace *colMustard = [[GameBoardSpace alloc] initWithId:ColMustardID];
    [[colMustard navigationTargets] addObject:[spaces valueForKey:@"Lounge-DiningRoom"]];
    [self addGameBoardSpace:colMustard];
    
    GameBoardSpace *mrsWhite = [[GameBoardSpace alloc] initWithId:MrsWhiteID];
    [[mrsWhite navigationTargets] addObject:[spaces valueForKey:@"BallRoom-Kitchen"]];
    [self addGameBoardSpace:mrsWhite];
    
    GameBoardSpace *mrGreen = [[GameBoardSpace alloc] initWithId:MrGreenID];
    [[mrGreen navigationTargets] addObject:[spaces valueForKey:@"Conservatory-BallRoom"]];
    [self addGameBoardSpace:mrGreen];
    
    GameBoardSpace *mrsPeacock = [[GameBoardSpace alloc] initWithId:MrsPeacockID];
    [[mrsPeacock navigationTargets] addObject:[spaces valueForKey:@"Library-Conservatory"]];
    [self addGameBoardSpace:mrsPeacock];
    
    GameBoardSpace *profPlum = [[GameBoardSpace alloc] initWithId:ProfPlumID];
    [[profPlum navigationTargets] addObject:[spaces valueForKey:@"Library-Study"]];
    [self addGameBoardSpace:profPlum];
}

-(void) addGameBoardSpace: (GameBoardSpace*) space
{
    [self.spaces setValue:space forKey:space.spaceId];
}

@end

File: src/src/edu/jhu/ep/butlerdidit/domain/GameBoard.java
package edu.jhu.ep.butlerdidit.domain;

import org.apache.http.impl.io.ContentLengthInputStream;

import java.util.*;

public class GameBoard {

    private List<Player> players;

    private Dictionary<String, GameBoardSpace> spaces;
    public Dictionary<String, GameBoardSpace> getSpaces() {
        return spaces;
    }

    public GameBoard(List<Player> players) {
        this.players = players;

        createSpaces();
    }

    public List<GameBoardSpace> getPossibleMoves(Player player) {
        GameBoardSpace currentSpace = spaces.get(player.getLocation());
        List<GameBoardSpace> possibleMoves = new Vector<GameBoardSpace>();

        for(GameBoardSpace space : currentSpace.getNavigationTargets()) {
            if(space instanceof HallwaySpace) {
                boolean isOccupied = false;
                for(Player otherPlayer : players) {
                    if(otherPlayer.getLocation().equals(space.getSpaceId())) {
                        isOccupied = true;
                        break;
                    }
                }
                if(!isOccupied) possibleMoves.add(space);
            } else {
                possibleMoves.add(space);
            }
        }
        return possibleMoves;
    }

    public boolean isPlayerAbleToMoveToSpace(Player player, GameBoardSpace space) {
        List<GameBoardSpace> possibleMoves = getPossibleMoves(player);

        for(GameBoardSpace target : possibleMoves) {
            if(target.getSpaceId() == space.getSpaceId()) return true;
        }
        return false;
    }

    private void createSpaces() {
        spaces = new Hashtable<String, GameBoardSpace>();

        Room study = new Room(Room.STUDY);
        Room hall = new Room(Room.HALL);
        HallwaySpace hallway = new HallwaySpace(study, hall);
        addSpaceToBoard(study);
        addSpaceToBoard(hall);
        addSpaceToBoard(hallway);

        Room lounge = new Room(Room.LOUNGE);
        hallway = new HallwaySpace(hall, lounge);
        addSpaceToBoard(lounge);
        addSpaceToBoard(hallway);

        Room diningRoom = new Room(Room.DININGROOM);
        hallway = new HallwaySpace(lounge, diningRoom);
        addSpaceToBoard(diningRoom);
        addSpaceToBoard(hallway);

        Room kitchen = new Room(Room.KITCHEN);
        hallway = new HallwaySpace(diningRoom, kitchen);
        addSpaceToBoard(kitchen);
        addSpaceToBoard(hallway);

        Room ballroom = new Room(Room.BALLROOM);
        hallway = new HallwaySpace(kitchen, ballroom);
        addSpaceToBoard(ballroom);
        addSpaceToBoard(hallway);

        Room conservatory = new Room(Room.CONSERVATORY);
        hallway = new HallwaySpace(ballroom, conservatory);
        addSpaceToBoard(conservatory);
        addSpaceToBoard(hallway);

        Room library = new Room(Room.LIBRARY);
        hallway = new HallwaySpace(conservatory, library);
        addSpaceToBoard(library);
        addSpaceToBoard(hallway);
        hallway = new HallwaySpace(library, study);
        addSpaceToBoard(hallway);

        Room billiardRoom = new Room(Room.BILLIARDROOM);
        addSpaceToBoard(billiardRoom);
        addSpaceToBoard(new HallwaySpace(billiardRoom, library));
        addSpaceToBoard(new HallwaySpace(billiardRoom, hall));
        addSpaceToBoard(new HallwaySpace(billiardRoom, diningRoom));
        addSpaceToBoard(new HallwaySpace(billiardRoom, ballroom));

        // Connect secret passages
        conservatory.connectSpace(lounge);
        study.connectSpace(kitchen);

        // Create start spaces
        GameBoardSpace msScarlet = new GameBoardSpace(ClueCharacter.MsScarletID);
        msScarlet.connectSpace(spaces.get("Hall-Lounge"));
        addSpaceToBoard(msScarlet);

        GameBoardSpace colMustard = new GameBoardSpace(ClueCharacter.ColMustardID);
        colMustard.connectSpace(spaces.get("Lounge-Dining Room"));
        addSpaceToBoard(colMustard);

        GameBoardSpace mrsWhite = new GameBoardSpace(ClueCharacter.MrsWhiteID);
        mrsWhite.connectSpace(spaces.get("Ball Room-Kitchen"));
        addSpaceToBoard(mrsWhite);

        GameBoardSpace mrGreen = new GameBoardSpace(ClueCharacter.MrGreenID);
        mrGreen.connectSpace(spaces.get("Conservatory-Ball Room"));
        addSpaceToBoard(mrGreen);

        GameBoardSpace mrsPeacock = new GameBoardSpace(ClueCharacter.MrsPeacockID);
        mrsPeacock.connectSpace(spaces.get("Conservatory-Library"));
        addSpaceToBoard(mrsPeacock);

        GameBoardSpace profPlum = new GameBoardSpace(ClueCharacter.ProfPlumID);
        profPlum.connectSpace(spaces.get("Library-Study"));
        addSpaceToBoard(profPlum);
    }

    private void addSpaceToBoard(GameBoardSpace space) {
        spaces.put(space.getSpaceId(), space);
    }
}
